### ISSUE TITLE
Unsiged Types Suffix on variable declaration

### ISSUE THREAD

**Issue by JustNrik** [dotnet/vblang#439]
Just like
```vb
Dim int%
```
is inferred as Integer, the suffix `%u` should make it UInteger as Unsigned Types currently don't benefit from this hidden treasure.

**Comment by pricerc** [comment]
Interesting concept. 

I never declare variables in this way, because I adhere to the philosophy that variable names should describe their 'use', not their 'type', and 'type characters' force the 'type' into the variable name. 

So how does it work if you decide that your `int%` needs to grow into a `long&` ?

**Comment by JustNrik** [comment]
> So how does it work if you decide that your `int%` needs to grow into a `long&` ?

I'm not sure exactly what you mean, a variable declared as integer won't mutate into long. `%` is just translated as ` As Integer` during compilation, my suggestion is to add the unsiged variants, which would be `%u` for UInteger and `&u` for ULong

**Comment by pricerc** [comment]
> > So how does it work if you decide that your `int%` needs to grow into a `long&` ?
> 
> I'm not sure exactly what you mean, a variable declared as integer won't mutate into long. `%` is just translated as ` As Integer` during compilation, my suggestion is to add the unsiged variants, which would be `%u` for UInteger and `&u` for ULong

Sorry, I wasn't clear. 

I understand you suggestion, what I don't understand, is why you want to use type characters. I.e. I'm trying to understand your programming methods.

Let me rephrase my question. Say I have a variable `id`. I write my code, and 32 bits is big enough, so:
```vbnet
Dim id As Integer
```
In a month's time, I discover that I was wrong. So I change my code to:
```vbnet
Dim id As Long
```

My variable's data type has changed, but the name is the same, so I don't need to change anything else.

In your case,
```vbnet
Dim id%
```
becomes
```vbnet
Dim id&
```
which you then have to change everywhere in my code.

Am I correct?

**Comment by AdamSpeight2008** [comment]
I'd probably use a type suffix `Dim i = 123UI`, `Dim i = 123UL`

**Comment by JustNrik** [comment]
> 
> 
> > > So how does it work if you decide that your `int%` needs to grow into a `long&` ?
> > 
> > 
> > I'm not sure exactly what you mean, a variable declared as integer won't mutate into long. `%` is just translated as ` As Integer` during compilation, my suggestion is to add the unsiged variants, which would be `%u` for UInteger and `&u` for ULong
> 
> Sorry, I wasn't clear.
> 
> I understand you suggestion, what I don't understand, is why you want to use type characters. I.e. I'm trying to understand your programming methods.
> 
> Let me rephrase my question. Say I have a variable `id`. I write my code, and 32 bits is big enough, so:
> 
> ```vbnet
> Dim id As Integer
> ```
> 
> In a month's time, I discover that I was wrong. So I change my code to:
> 
> ```vbnet
> Dim id As Long
> ```
> 
> My variable's data type has changed, but the name is the same, so I don't need to change anything else.
> 
> In your case,
> 
> ```vbnet
> Dim id%
> ```
> 
> becomes
> 
> ```vbnet
> Dim id&
> ```
> 
> which you then have to change everywhere in my code.
> 
> Am I correct?

No. This is a feature of VB called "Identifier type characters", they are compile-time suffixes that tells the compiler the type of the variable declared, these suffixes are just a compile-thing. You can write your code normally with the name of the variable without the suffix.

```vb
Dim value%
' It gets compiled into
Dim value As Integer
' You use it as "value", not as "value%"
' If you upgrade it to long, % is replaced with & but the name of the variable is the same.
```



> 
> 
> I'd probably use a type suffix `Dim i = 123UI`, `Dim i = 123UL`

I'm talking about `Identifier type characters`, not `Literal type characters`, the purpose of my suggestion is to extend this barely known feature for unsigned types (well, UInteger and ULong specifically)

More info there -> https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/language-features/data-types/type-characters

**Comment by AdamSpeight2008** [comment]
If it barely known about, does it indicate that it could rarely used?
Have you looked into how often those `Identifier Type Characters` are used?

**Comment by JustNrik** [comment]
> 
> 
> If it barely known about, does it indicate that it could rarely used?
> Have you looked into how often those `Identifier Type Characters` are used?

Not much, I literally discovered them really recently looking at some legacy projects in my college (vs 2008 and older projects). At least myself I find them actually useful to reduce verbosity, specially on method arguments, for example
```vb
Function DoSomething(str1$, str2$) As Type
    ' Do something with str1 and str2
    Return result
End Function
```
looks cleaner than adding `As String, As String` IMHO. Maybe it won't be such a used feature but I guess the few people that would actually benefit from it will be grateful, after all, VB is a language that is designed to give as much facilities as possible, this is one of these facilities.

**Comment by pricerc** [comment]
@JustNrik Actually, type characters come from the beginnings of BASIC, when that was the only way to declare the type of a variable, and there was no such concept as a class or a function.

I believe that the style is now considered 'legacy', and to be discouraged; they are really only still there for backward compatibility with ancient code, so I would be surprised if this proposal got any traction from the LDM.

'Cleaner' is completely subjective, and in this case, I'd disagree with you, because I expect to see an 'As <Type>' clause on each parameter declaration, and I find not having them to be 'disruptive'; like when you're reading a sentence that doesn't have

**Comment by pricerc** [comment]
an ending

**Comment by VBAndCs** [comment]
I don't like these type symbols, and never used them. In fact they cause troubles, for example, prevent me from using the @ symbol as a var name or a part of any var name, because it is preserved for some type I can't remember! This legacy way is discarded as it makes code less readable.

**Comment by DualBrain** [comment]
I'm actually a fan of using these... I've experimented with projects where I use these exclusively (where I've done public presentations on).  Once you get used to it, it's actually very easy to quickly type:

Dim x%

x += 1

However, with that said, I've also spent a lot of time using Option Infer On and really like that approach.  However, the % isn't just for the variable... it's also shorthand to "force" a value type.  For example, you can save yourself a lot of casting when you have to add numbers:

Dim x = 0UI

It looks a little strange... but that actually will accomplish the same goal of what was originally asked.  So one could argue that we basically already have this.

And, please, for the love of all that is holy, I wish people would stop referring to things as "legacy" and "discouraged".  Just because you (and a few of your friends) think something is "icky" doesn't mean that there aren't many others out there who think it's the best thing since sliced bread.  There are merits to all the approaches and having choice is what makes VB such a wonderful language.

**Comment by DualBrain** [comment]
Also, @ is a new one that was added as part of VB.NET for the Decimal type... just saying. ;-)
